### ISSUE TITLE
Cannot install extension on remote server

### ISSUE THREAD

**Issue by cross-cisco** [microsoft/vscode-remote-release#10571]
Type: <b>Bug</b>

I am running VS Code 1.96.0 on a Mac, MacOS Sequoia 15.1.1, Darwin 24.1.0.
I have installed the Remote Development and Remote-SSH extensions, which work well for allowing me to edit files on a remote linux server.  However, when I open the extensions panel, it shows no extensions installed on the server, and if I try to install an extension on the remote using the button on the info panel for the extension, it reports "Installing" forever.

The javascript console is reporting errors about non-existant files (on my Mac) inside of User/globalStorage that appear not to exist.  I'm not sure why it's looking for extension data here.

But, assuming that error message is not the problem, there is some other problem I'm not able to see.  The "Remote - SSH" Output view shows the extension copied into data/CachedExetnsionVSIXs/UUID, and then reports "Installing extension:" and a JSON block.  But, then it just sits.

Extension version: 0.116.1
VS Code version: Code 1.96.0 (Universal) (138f619c86f1199955d53b4166bef66ef252935c, 2024-12-11T02:29:09.626Z)
OS version: Darwin x64 24.1.0
Modes:
Remote OS version: Linux x64 6.8.0-49-generic

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz (16 x 2400)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|5, 4, 4|
|Memory (System)|64.00GB (9.91GB free)|
|Process Argv|--crash-reporter-id 88d235c0-e2c1-4cbe-ba66-704cda98c861|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|SSH: aced-build2.cisco.com|
|OS|Linux x64 6.8.0-49-generic|
|CPUs|Intel(R) Xeon(R) CPU E5-2670 v3 @ 2.30GHz (48 x 1200)|
|Memory (System)|377.77GB (370.69GB free)|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscod805cf:30301675
binariesv615:30325510
vsaa593:30376534
py29gd2263:31024239
vscaat:30438848
c4g48928:30535728
azure-dev_surveyone:30548225
a9j8j154:30646983
962ge761:30959799
pythonnoceb:30805159
pythonmypyd1:30879173
h48ei257:31000450
pythontbext0:30879054
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
dsvsc021:30996838
dvdeprecation:31068756
dwnewjupytercf:31046870
newcmakeconfigv2:31071590
nativerepl1:31139838
pythonrstrctxt:31112756
nativeloc1:31192215
cf971741:31144450
iacca1:31171482
notype1:31157159
5fd0e150:31155592
dwcopilot:31170013
stablechunks:31184530
6074i472:31201624

```

</details>

**Comment by joshspicer** [comment]
Thank you for taking the time to open an issue.  Could you please share your logs (`> Remote-SSH: Show Log` in the command palette to open in the Output panel) after reproducing the issue? Also please share your "remote extension host" and "window" logs from the same Output panel

**Comment by cross-cisco** [comment]
Thank you.  Okay, I opened VS Code with the prior window being a remote-ssh to the given host.  It establishes the connection, many processes are running on the server, and the client seems idle.
A few minutes later, I clicked on the "Extensions" icon in the left sidebar.  The sidebar showed local extensions, and an empty divider for the server with 0 extensions.  There was an ongoing cylon at the top of that side panel.  I'm guessing it's unable to get the list of extensions from the server and is trying at that point?  After about a minute of watching it do nothing, I dumped the logs.

Second document is the two Output windows you requested.

[remote-ssh-show-log.txt](https://github.com/user-attachments/files/18170142/remote-ssh-show-log.txt)
[Output-windows.txt](https://github.com/user-attachments/files/18170143/Output-windows.txt)

**Comment by cross-cisco** [comment]
Oh.  :-/  `Module_GetConnectionEvents_Test` tests that empty files get removed.  My earlier change modified that behaviour only for `Module_GetConnectionEventsTruncated_Test` but now that it's across the system it breaks that test.
Happy to hear suggestions, I'll get back to it later.

**Comment by cross-cisco** [comment]
Is there anything else I can provide?  Any suggestions for a workaround?  I would like to use Copilot with my remote-ssh project(s).

**Comment by joshspicer** [comment]
Do these workaround steps work?  https://github.com/microsoft/vscode-remote-release/issues/10578

**Comment by cross-cisco** [comment]
> Do these workaround steps work? [#10578](https://github.com/microsoft/vscode-remote-release/issues/10578)

No.  I don't see any extensions in the remote, so I can't disable or reenable them as that workaround specific to Copilot Chat suggests.

**Comment by joshspicer** [comment]
Ok, thanks for looking. I saw "Copilot" in your response but I see now that wouldn't have helped in your scenario.

Do you have another SSH remote (maybe on your home network?) you can try to see if the issue repros? Is your remote a work machine that may have some firewall or network restrictions in place? 

We might get more logging in the Remote - SHH output channel if you add an extension to the 'defaultExtensions' setting.  Could you try that?

Finally, mind having a look in the other Output channel logs and sharing anything that could look related? In the command palette you can boost your "log level" up to "trace" so that more is emitted.  I expected "remote extension host" or "window" to be the right places, but perhaps the issue could be found elsewhere?

cc/ @deepak1556 or @connor4312, do you have any tips you could share with me on debugging this?

**Comment by connor4312** [comment]
There's an existing bug (fixed in the release ~10 minutes from now) that fixes copilot not working on remotes https://github.com/microsoft/vscode/issues/236537

**Comment by joshspicer** [comment]
@connor4312 from the sounds of what @cross-cisco is sharing this isn't the same issue.

That said, @cross-cisco please do try out the new build and let us know in case it was a different repro with the same root cause 🙏

**Comment by cross-cisco** [comment]
It's not specific to copilot.  There are no extensions on the remote, according to my local VSCode, and I can't install any.
On a related note, a new update to VS Code seems to be available, and this same remote server is having no luck getting it.  It's just been saying "Downloading VS Code Server..." for 10 minutes now, and I can see an empty `cli/servers/Stable-fabdb6a30b49f79a7aba0f2ad9df9b399473380f.staging` in my `.vscode-server` space.

Let me know where I can get data to help.  nb, I see
`[16:16:09.413] [server] Installing and setting up Visual Studio Code Server...`
in the "Remote - SSH" Output, and I see multiple occurrences of
`2024-12-19 16:16:08.593 [warning] Shell integration cannot be enabled for executable "/bin/sh" ...`
in the Pty Host output window.  However the last one of these (above) is for the "connected, do not close" then print a space every 3 minutes.  And, I can see that message and the incoming space characters in my Terminal pane.  So it is clearly running despite the "shell integration cannot be enabled" message.

**Comment by joshspicer** [comment]
Thanks for bearing with us.  Could you check the output from the 'Server' tab after attempting to install an extension.  For example, this is what I see when connected to a remote and installing a random extension:

<img width="1005" alt="Image" src="https://github.com/user-attachments/assets/e274a8c8-e57e-4abf-baab-efd36502b257" />

**Comment by cross-cisco** [comment]
Right now I cannot.  I am stuck at the other, possibly related, issue where the update to Visual Studio code won't install.  (I also don't see "Server" in my output list, but that may just be because I'm earlier in the process.)
![Image](https://github.com/user-attachments/assets/fa9b5e3d-d88f-477d-beab-e943b93affcf)

**Comment by cross-cisco** [comment]
> Right now I cannot. I am stuck at the other, possibly related, issue where the update to Visual Studio code won't install. 

I have in the past found that I can copy the updated VSCode from another system where it _will_ install correctly.  But I don't want to use that workaround and bypass the opportunity to diagnose this problem.

**Comment by cross-cisco** [comment]
Okay.  I've turned off "Exec Server" in the Remote SSH settings, and now it is able to update the server installion on the remote host.  However, I still see no extensions listed, and installing one continues to just appear stuck.
Let me know what to try next @joshspicer

**Comment by cross-cisco** [comment]
Let me know what information I can provide. I am able to use the remote-ssh connection to my server, however it still shows no extensions installed.  I have a little clock-face icon laid over the extensions icon in the far left, and an ingoing cylon at the top of the sidebar.

**Comment by joshspicer** [comment]
Apologies for the delay, I understand you want this resolved.  @deepak1556 perhaps you have some idea of how to capture extension installation logs?  @cross-cisco is sharing that they do not have the 'server' tab in their output window, and i'm not sure why that is

**Comment by cross-cisco** [comment]
I'm not sure why I didn't see it earlier, but I do have "Server" output in "OUTPUT" now.  Attached.

![Image](https://github.com/user-attachments/assets/391192da-212d-43cd-ba2e-93323124b2c6)

The files mentioned in the above are as follows:
```
cross@aced-build2 /data/cr/p/srcdir ❯ ls -lA /users/cross2/.vscode-server/data/CachedExtensionVSIXs/4404961b-478a-4324-a8b9-f7adf876f69a
-rw-r--r-- 1 cross2 eng 15199501 Jan 21 11:08 /users/cross2/.vscode-server/data/CachedExtensionVSIXs/4404961b-478a-4324-a8b9-f7adf876f69a
cross@aced-build2 /data/cr/p/srcdir ❯ ls -l /users/cross2/.vscode-server/extensions/extensions.json
-rw-r--r-- 1 cross2 eng 1426 Oct 31 11:00 /users/cross2/.vscode-server/extensions/extensions.json
cross@aced-build2 /data/cr/p/srcdir ❯ 
```

Noting:
- No extensions listed on the remote server
- The mark in the bar under "EXTENSIONS" at the top of the left sidebar continues to cylon, moving left-to-right forever.

**Comment by cross-cisco** [comment]
Checking in.  I am still seeing this same behavior.  When I open the remote SSH connection to my Linux servers, even before opening any directories to look at a project, the EXTENSIONS tab shows an ever-moving cylon at the top, locally installed extensions (LOCAL - INSTALLED), and nothing in either the "SSH: servername - INSTALLED" or "RECOMMENDED" sections.

What else can I try?

**Comment by cross-cisco** [comment]
Hmm.  Looking through the various output windows, I see a couple of things
1. There are at least _some_ extensions on the remote server because the "Extension Host (Remote)" output window contains:
```
2025-01-30 16:10:25.603 [info] Extension host with pid 2836256 started
2025-01-30 16:10:25.723 [info] ExtensionService#_doActivateExtension vscode.git-base, startup: true, activationEvent: '*', root cause: vscode.git
2025-01-30 16:10:25.725 [info] ExtensionService#_doActivateExtension mjcrouch.perforce, startup: true, activationEvent: '*'
2025-01-30 16:10:25.874 [info] ExtensionService#_doActivateExtension vscode.git, startup: true, activationEvent: '*'
2025-01-30 16:10:25.874 [info] ExtensionService#_doActivateExtension vscode.github, startup: true, activationEvent: '*'
2025-01-30 16:10:27.056 [info] Eager extensions activated
2025-01-30 16:10:27.056 [info] ExtensionService#_doActivateExtension vscode.debug-auto-launch, startup: false, activationEvent: 'onStartupFinished'
2025-01-30 16:10:27.057 [info] ExtensionService#_doActivateExtension vscode.merge-conflict, startup: false, activationEvent: 'onStartupFinished'
2025-01-30 16:10:27.935 [info] ExtensionService#_doActivateExtension vscode.emmet, startup: false, activationEvent: 'onLanguage'
```
2. There are a bunch of `2025-01-30 16:10:23.238 [warning] Shell integration cannot be enabled for executable "/bin/sh" and args ["-c","wget --version > /dev/null\nif [ $? -eq 0 ]\nthen\n\twget --no-config --`[...] type of messages in "Pty Host (Remote)" console. I'm not sure if those are normal?  They do all seem to be "metadata from cloud service" type requests and my server is not in a cloud.
3. "Main" console shows the following.  Is that last **[error]** a real problem?
```
2025-01-30 16:08:19.670 [info] update#setState idle
2025-01-30 16:08:45.877 [info] Extension host with pid 22822 exited with code: 0, signal: unknown.
2025-01-30 16:08:49.442 [info] update#setState checking for updates
2025-01-30 16:08:49.745 [info] update#setState idle
2025-01-30 16:10:07.173 [info] Extension host with pid 23450 exited with code: 0, signal: unknown.
2025-01-30 16:13:55.243 [info] Extension host with pid 23008 exited with code: 0, signal: unknown.
2025-01-30 16:13:55.388 [error] Blocked vscode-file request vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/workbench/contrib/output/common/outputLinkComputerMain.js#OutputLinkDetectionWorker
```
4. "Server" output text shows the following.  (It's now 16:24 in my local time)  I find that 5-minute gap interesting.  Maybe not important, but I thought it was worth noting.

```
2025-01-30 16:10:15.605 [info] 




2025-01-30 16:10:15.852 [info] Extension host agent started.
2025-01-30 16:10:17.009 [info] [<unknown>][07c90ac5][ManagementConnection] New connection established.
2025-01-30 16:10:17.046 [info] [<unknown>][893b8f06][ExtensionHostConnection] New connection established.
2025-01-30 16:10:23.002 [info] [<unknown>][893b8f06][ExtensionHostConnection] <2836256> Launched Extension Host Process.
2025-01-30 16:15:15.853 [info] New EH opened, aborting shutdown
```

Anyway.  As noted, this is all just on connect without opening a folder on the remote.  Most of what I'd recorded in earlier comments/data was after looking into my workspaces.  This all shows up just after connecting, so it's not likely related to my workspaces on the remote.

Thanks for your efforts.

**Comment by cross-cisco** [comment]
Hmm.   An interesting note.  If I use Remote-SSH to connect to a Raspberry Pi, I can see remote extensions.  However, I do see the following, which repeats a minute or more later after selecting "Restart" each time.

![Image](https://github.com/user-attachments/assets/3843886e-047b-4b99-9621-fc65d37a2ecd)

I'm still having the same problem with my Ubuntu and RHEL hosts showing no extensions.  Currently, the vscode on the Ubuntu host(s) is unable to update, and it's stuck there.  As mentioned above, I think there are comms issues affecting both of these.  Any help in diagnosing would be appreciated.

**Comment by cross-cisco** [comment]
@joshspicer is there another issue open on the "fail to update remote code installation using Remote SSH" ?  If so we can leave this issue related to the original "remote extensions not visible" even if it's the same underlying problem.

I have other users here now suffering that same issue on these same machines.  I would love some insight or suggestions.  :-)

**Comment by cross-cisco** [comment]
tl;dr; I think I have a simple test which demonstrates the problem...


Hello again.  After talking with a coworker today, I chose to try something.  (On one of the servers I am having this problem on, I created a new local user.  Most users have NFS home direcories.  In my case, I created a `~/.vscode-server` to a local disk directory, but it's still an atypical case.)

I've created a new user, local disk home directory.  The first time I try to connect to the server with vscode "Remote-SSH: Connect current window to host", it creates ~/.vscode-server, and fails to download the server.  I'd mentioned above that this sometimes fails.

Terminal window shows:
```
cross2local@aced-build3.cisco.com's password: 
9fd0a4237440: running
Acquiring lock on /home/cross2local/.vscode-server/bin/e54c774e0add60467559eb0d1e229c6452cf8447/vscode-remote-lock.cross2local.e54c774e0add60467559eb0d1e229c6452cf8447
Installing to /home/cross2local/.vscode-server/bin/e54c774e0add60467559eb0d1e229c6452cf8447...
9fd0a4237440%%1%%
Downloading with wget

```

And then it just sits.  Modal showing "Setting up SSH Host ...: Downloading VS Code".  On the server side, I can see the wget running, but its output file is 0 bytes.

If I try running the same wget with `--debug`, I see:
```
❯ wget --no-config --verbose --debug --tries=1 --connect-timeout=7 --dns-timeout=7 -nv -O vscode-server.tar.gz https://update.code.visualstudio.com/commit:e54c774e0add60467559eb0d1e229c6452cf8447/server-linux-x64/stable
Setting --tries (tries) to 1
Setting --connect-timeout (connecttimeout) to 7
Setting --dns-timeout (dnstimeout) to 7
Setting --no (verbose) to 0
Setting --output-document (outputdocument) to vscode-server.tar.gz
DEBUG output created by Wget 1.21.4 on linux-gnu.

Reading HSTS entries from /home/cross2local/.wget-hsts
URI encoding = ‘UTF-8’
Caching update.code.visualstudio.com => 2620:1ec:bdf::57 13.107.246.57
Created socket 4.
Releasing 0x000057bc6e0900d0 (new refcount 1).
Initiating SSL handshake.


```
and then it just sits there, doing nothing until I hit Ctrl-C.

Let me know if this can provide any insight.  And I have a wget command that shows the problem easily, so woot!  :-)

**Comment by cross-cisco** [comment]
\*f\*ck\*. It's an IPv6 problem.  If I add a "-4" to the wget command line, it all works fine.

So, at least the failure to load a new vscode isn't software related.  There's something wrong with IPv6 on microsoft's side.

```❯ ping6 update.code.visualstudio.com
PING update.code.visualstudio.com (2620:1ec:bdf::57) 56 data bytes
^C
--- update.code.visualstudio.com ping statistics ---
5 packets transmitted, 0 received, 100% packet loss, time 4112ms

❯ ping6 www.google.com
PING www.google.com (2607:f8b0:4005:80f::2004) 56 data bytes
64 bytes from nuq04s42-in-x04.1e100.net (2607:f8b0:4005:80f::2004): icmp_seq=1 ttl=105 time=94.9 ms
64 bytes from nuq04s42-in-x04.1e100.net (2607:f8b0:4005:80f::2004): icmp_seq=2 ttl=105 time=92.5 ms
^C
--- www.google.com ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1002ms
rtt min/avg/max/mdev = 92.541/93.700/94.859/1.159 ms
```

But it's not raw connectivity, ping not relevant.  The TCP connection to update.code.visualstudio.com via IPv6 is successful.  It then just fails to send or receive the traffic it seems.  Later testing with `curl -v` shows that the TLS handshake seems to exchange bytes, so it's later than that it stops.

@joshspicer this here is _not_ a `vscode-remote` issue, for sure.  Though let's leave this open until we rule this out as a cause for that too later in the process.
Can you find someone more on the networking side, and point me to a conversation elsewhere?

**Comment by joshspicer** [comment]
Can you reach our extensions CDN?

```
curl https://main.vscode-cdn.net/extensions/marketplace.json --verbose 1> /dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Host main.vscode-cdn.net:443 was resolved.
* IPv6: 2620:1ec:29:1::70
* IPv4: 13.107.253.70
*   Trying [2620:1ec:29:1::70]:443...
*   Trying 13.107.253.70:443...
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to main.vscode-cdn.net (2620:1ec:29:1::70) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
} [324 bytes data]
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
{ [88 bytes data]
* (304) (OUT), TLS handshake, Client hello (1):
} [357 bytes data]
* (304) (IN), TLS handshake, Server hello (2):
{ [155 bytes data]
* (304) (IN), TLS handshake, Unknown (8):
{ [19 bytes data]
* (304) (IN), TLS handshake, Certificate (11):
{ [4548 bytes data]
* (304) (IN), TLS handshake, CERT verify (15):
{ [264 bytes data]
* (304) (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* (304) (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / AEAD-AES256-GCM-SHA384 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: C=US; ST=WA; L=Redmond; O=Microsoft Corporation; CN=*.vscode-cdn.net
*  start date: Feb 10 09:07:25 2025 GMT
*  expire date: Aug  9 09:07:25 2025 GMT
*  subjectAltName: host "main.vscode-cdn.net" matched cert's "*.vscode-cdn.net"
*  issuer: C=US; O=Microsoft Corporation; CN=Microsoft Azure RSA TLS Issuing CA 08
*  SSL certificate verify ok.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://main.vscode-cdn.net/extensions/marketplace.json
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: main.vscode-cdn.net]
* [HTTP/2] [1] [:path: /extensions/marketplace.json]
* [HTTP/2] [1] [user-agent: curl/8.7.1]
* [HTTP/2] [1] [accept: */*]
> GET /extensions/marketplace.json HTTP/2
> Host: main.vscode-cdn.net
> User-Agent: curl/8.7.1
> Accept: */*
>
* Request completely sent off
< HTTP/2 200
< date: Wed, 26 Feb 2025 21:28:59 GMT
< content-type: application/json
< content-length: 50357
< cache-control: max-age=3600
< last-modified: Wed, 26 Feb 2025 13:59:42 GMT
< etag: "0x8DD566DD16165CE"
< x-ms-request-id: ....
< x-ms-version: 2018-03-28
< x-azure-ref: ...
< x-fd-int-roxy-purgeid: 3
< x-cache: TCP_HIT
< vary: Origin
< cross-origin-resource-policy: cross-origin
< accept-ranges: bytes
<
{ [8192 bytes data]
100 50357  100 50357    0     0  44551      0  0:00:01  0:00:01 --:--:-- 44524
* Connection #0 to host main.vscode-cdn.net left intact
```

**Comment by chrmarti** [comment]
@cross-cisco Thanks for helping investigate this! I was just trying this myself and it seems to work for me with ipv6. Could this be a network infrastructure issue? Have you tried with `-6` to rule out any ipv6 vs. ipv4 fallback issues? Anything I can try on my side to compare with? I'm not a network infrastructure person, but I'm happy to find the right one once we have narrowed the issue down further.

Seeing:
```
wget --no-config --verbose --debug --tries=1 -O vscode-server.tar.gz https://update.code.visualstudio.com/commit:e54c774e0add60467559eb0d1e229c6452cf8447/server-linux-x64/stable 

Setting --tries (tries) to 1
Setting --output-document (outputdocument) to vscode-server.tar.gz
DEBUG output created by Wget 1.24.5 on darwin23.2.0.

Reading HSTS entries from /Users/chrmarti/.wget-hsts
URI encoding = ‘UTF-8’
--2025-02-27 08:31:41--  https://update.code.visualstudio.com/commit:e54c774e0add60467559eb0d1e229c6452cf8447/server-linux-x64/stable
Resolving update.code.visualstudio.com (update.code.visualstudio.com)... 2620:1ec:bdf::60, 13.107.246.60
Caching update.code.visualstudio.com => 2620:1ec:bdf::60 13.107.246.60
Connecting to update.code.visualstudio.com (update.code.visualstudio.com)|2620:1ec:bdf::60|:443... connected.
Created socket 6.
Releasing 0x0000000146752c70 (new refcount 1).
Initiating SSL handshake.
Handshake successful; connected socket 6 to SSL handle 0x0000000147008200
certificate:
  subject: CN=update.code.visualstudio.com,O=Microsoft Corporation,L=Redmond,ST=WA,C=US
  issuer:  CN=Microsoft Azure RSA TLS Issuing CA 07,O=Microsoft Corporation,C=US
X509 certificate successfully verified and matches host update.code.visualstudio.com

---request begin---
GET /commit:e54c774e0add60467559eb0d1e229c6452cf8447/server-linux-x64/stable HTTP/1.1
Host: update.code.visualstudio.com
User-Agent: Wget/1.24.5
Accept: */*
Accept-Encoding: identity
Connection: Keep-Alive

---request end---
HTTP request sent, awaiting response... 
---response begin---
HTTP/1.1 302 Found
Date: Thu, 27 Feb 2025 07:31:41 GMT
Content-Type: text/plain; charset=utf-8
Content-Length: 160
Connection: keep-alive
Access-Control-Allow-Origin: *
Cache-Control: max-age=150
Location: https://vscode.download.prss.microsoft.com/dbazure/download/stable/e54c774e0add60467559eb0d1e229c6452cf8447/vscode-server-linux-x64.tar.gz
Vary: Accept
Strict-Transport-Security: max-age=31536000; includeSubDomains
request-context: appId=cid-v1:8109e553-6232-492a-afca-b78cea9b431c
X-Powered-By: Express
X-Content-Type-Options: nosniff
X-Source-Commit: 52c757d6039fc7b9fc4b210aec30978ae2805866
X-SHA256: 95608b83c284aa09a7e26b955219e958d35ca2bf58fb8b08bf9356ac0320cd95
x-azure-ref: 20250227T073141Z-15c7f56586f66dp8hC1ZRHw8t40000000fg0000000008w59
X-Cache: CONFIG_NOCACHE

---response end---
302 Found
Registered socket 6 for persistent reuse.
Parsed Strict-Transport-Security max-age = 31536000, includeSubDomains = true
Updated HSTS host: update.code.visualstudio.com:443 (max-age: 31536000, includeSubdomains: true)
URI content encoding = ‘utf-8’
Location: https://vscode.download.prss.microsoft.com/dbazure/download/stable/e54c774e0add60467559eb0d1e229c6452cf8447/vscode-server-linux-x64.tar.gz [following]
Skipping 160 bytes of body: [Found. Redirecting to https://vscode.download.prss.microsoft.com/dbazure/download/stable/e54c774e0add60467559eb0d1e229c6452cf8447/vscode-server-linux-x64.tar.gz] done.
URI content encoding = None
--2025-02-27 08:31:41--  https://vscode.download.prss.microsoft.com/dbazure/download/stable/e54c774e0add60467559eb0d1e229c6452cf8447/vscode-server-linux-x64.tar.gz
Resolving vscode.download.prss.microsoft.com (vscode.download.prss.microsoft.com)... 2001:918:ffad::c1f7:2b33, 2001:918:ffad::c1f7:2b18, 193.246.105.41, ...
Caching vscode.download.prss.microsoft.com => 2001:918:ffad::c1f7:2b33 2001:918:ffad::c1f7:2b18 193.246.105.41 193.246.105.65
Connecting to vscode.download.prss.microsoft.com (vscode.download.prss.microsoft.com)|2001:918:ffad::c1f7:2b33|:443... connected.
Created socket 7.
Releasing 0x0000000146711860 (new refcount 1).
Initiating SSL handshake.
Handshake successful; connected socket 7 to SSL handle 0x000000014700a000
certificate:
  subject: CN=download.prss.microsoft.com,O=Microsoft Corporation,L=Redmond,ST=WA,C=US
  issuer:  CN=Microsoft Azure ECC TLS Issuing CA 04,O=Microsoft Corporation,C=US
X509 certificate successfully verified and matches host vscode.download.prss.microsoft.com

---request begin---
GET /dbazure/download/stable/e54c774e0add60467559eb0d1e229c6452cf8447/vscode-server-linux-x64.tar.gz HTTP/1.1
Host: vscode.download.prss.microsoft.com
User-Agent: Wget/1.24.5
Accept: */*
Accept-Encoding: identity
Connection: Keep-Alive

---request end---
HTTP request sent, awaiting response... 
---response begin---
HTTP/1.1 200 OK
Content-Type: application/octet-stream
Server: Kestrel
Accept-Ranges: bytes
Cache-Control: public, max-age=259200
X-Ms-ApiVersion: Distribute 1.2
X-Ms-Region: prod-eus-z1
Content-Disposition: attachment; filename=vscode-server-linux-x64.tar.gz; filename*=UTF-8''vscode-server-linux-x64.tar.gz
Last-Modified: Thu, 13 Feb 2025 01:08:31 GMT
ETag: "0xF05EC7A0EE92D0A41D320BD25D2413D1E0B0AEC292E81BD467828242C091692C"
Content-Length: 61851007
Date: Thu, 27 Feb 2025 07:31:41 GMT
Connection: keep-alive
X-OSID: 2
X-CID: 2
X-CCC: CH

---response end---
200 OK
Disabling further reuse of socket 6.
Closed 6/SSL 0x0000000147008200
Registered socket 7 for persistent reuse.
Length: 61851007 (59M) [application/octet-stream]
Saving to: ‘vscode-server.tar.gz’

vscode-server.tar.gz                  100%[========================================================================>]  58.99M  47.6MB/s    in 1.2s    

2025-02-27 08:31:43 (47.6 MB/s) - ‘vscode-server.tar.gz’ saved [61851007/61851007]

Saving HSTS entries to /Users/chrmarti/.wget-hsts
```

**Comment by chrmarti** [comment]
Since the connection can be established, but the SSL handshake fails: Could this be an MTU issue? Is automatic discovery enabled (`sysctl net.inet.tcp.path_mtu_discovery`)?

**Comment by cross-cisco** [comment]
I did try it with both `-4` and `-6` with consistent results.  `-6` fails in the same way as not specifying.  It succeeds with IPv6 but doesn't _really_ succeed.

I do not have a inet-wide MTU discovery config. Ubuntu 24.04, kernel 6.8.0.  I do have all of the "accept mtu" options on however.
```
cross2@hostname /tmp ❯ host update.code.visualstudio.com | grep -v 'alias'
s-part-0029.t-0009.t-msedge.net has address 13.107.246.57
s-part-0029.t-0009.t-msedge.net has IPv6 address 2620:1ec:bdf::57
cross2@hostname /tmp ❯ ip -6 route get 2620:1ec:bdf::57                   
2620:1ec:bdf::57 from :: via 2001:420:2710:2556::1 dev enp2s0f0 proto static src 2001:420:2710:2556::105 metric 1024 pref medium
cross2@hostname /tmp ❯ sudo sysctl -a | grep '^net\.ipv6' | grep --color=no mtu
net.ipv6.conf.all.accept_ra_mtu = 1
net.ipv6.conf.all.mtu = 1280
net.ipv6.conf.default.accept_ra_mtu = 1
net.ipv6.conf.default.mtu = 1280
net.ipv6.conf.docker0.accept_ra_mtu = 1
net.ipv6.conf.docker0.mtu = 1500
net.ipv6.conf.eno1.accept_ra_mtu = 1
net.ipv6.conf.eno1.mtu = 1500
net.ipv6.conf.eno2.accept_ra_mtu = 1
net.ipv6.conf.eno2.mtu = 1500
net.ipv6.conf.enp2s0f0.accept_ra_mtu = 1
net.ipv6.conf.enp2s0f0.mtu = 1500
net.ipv6.conf.enp2s0f1.accept_ra_mtu = 1
net.ipv6.conf.enp2s0f1.mtu = 1500
net.ipv6.conf.lo.accept_ra_mtu = 1
net.ipv6.conf.lo.mtu = 65536
net.ipv6.conf.veth4936ed2.accept_ra_mtu = 1
net.ipv6.conf.veth4936ed2.mtu = 1500
net.ipv6.conf.vetha5e30e9.accept_ra_mtu = 1
net.ipv6.conf.vetha5e30e9.mtu = 1500
net.ipv6.route.mtu_expires = 600
cross2@hostname /tmp ❯ 
```

the relevant interface has a 1500 byte mtu it appears, but accepts MTU from the router.  Let me know if you think an MTU of 1500 is a problem, I can talk to the network folks about how IPv6 RA MTU is set up and/or working.

**Comment by chrmarti** [comment]
Not sure this makes sense, but to rule out issues with the MTU, you could reduce it to, e.g., 1000 on the interface and check if that changes anything? (RA MTU would only decrease the MTU I think?)

You could also use `openssl` to check if that can do the handshake: `openssl s_client -connect update.code.visualstudio.com:443 -6`.

**Comment by cross-cisco** [comment]
Okay.  So, now things have changed.  I did try that `openssl s_client` command last week, and it just emitted
```
CONNECTED(00000003)
```
and then just sat.

But, it works fine now.  I actually started today with wget, which also surprisingly worked fine.

I didn't get to trying to reset the mtu, and now things are working, so.  Let me see how VSCode behaves...
Today it loads alright, and seems to have downloaded `e54c774e0add60467559eb0d1e229c6452cf8447` onto the remote server.

Switching to a different server, it also loads `bin/e54c774e0add60467559eb0d1e229c6452cf8447` and I can see (and use) extensions on the remote server.

So, something must've changed in the networking between my side and yours (microsoft's cdn) over the last 4 days.  I'm not sure if this is a fixed issue or non-issue, or if this needs to be some sort of improvement in handling this issue.  It seems wget doesn't ever fall back to another address if it can connect but cannot complete the transfer.  I wonder if curl [mis]behaves the same way?

**Comment by chrmarti** [comment]
Have you tried with `-6`? It could be that ipv4 was slow to connect previously such that ipv6 was preferred and now ipv4 is again fast and is used by automatic selection.

I think it is up to each application to do the ipv4/6 selection (Happy Eyes) algorithm and that probably gives up on the alternative connection once a first connection is establish but before the SSL handshake happened.

You could install the Network Proxy Test extension (https://marketplace.visualstudio.com/items?itemName=chrmarti.network-proxy-test) and check the output of `F1` > `Network Proxy Test: Test Connection` for `https://update.code.visualstudio.com`. That has a DNS section with the available IP addresses. Not sure if that will help much though, maybe the extension could show which IP address it ended up using.

We have seen occasional issues with ipv6 in WSL, improvements to how we can diagnose and maybe counter these remain interesting.

**Comment by cross-cisco** [comment]
Oh, yes, sorry.  I did the external wget testing with `-6` and was surprised to see it work above.  I should've specified that.  And the `openssl s_client` command was with `-6` as originally specified.  It's definitely that the IPv6 connectivity which wasn't working properly has begun working.

**Comment by cross-cisco** [comment]
So, this has continued to work properly across recent weeks.  So, it is safe to assume that it's a MS infra issue with IPv6 connectivity, and that it has been addressed?
If that's known to be true, we can close this.  If there are still open questions, let's list them here to document the status.
